File: docker-gradle-plugin-main/src/main/java/com/github/godfather1103/gradle/tasks/AbstractDockerMojo.java
package com.github.godfather1103.gradle.tasks;

import com.github.dockerjava.api.DockerClient;
import com.github.dockerjava.core.DefaultDockerClientConfig;
import com.github.dockerjava.core.DockerClientConfig;
import com.github.dockerjava.core.DockerClientImpl;
import com.github.dockerjava.netty.NettyDockerCmdExecFactory;
import com.github.godfather1103.gradle.entity.AuthConfig;
import com.github.godfather1103.gradle.ext.DockerPluginExtension;
import com.google.common.base.Optional;
import org.apache.commons.lang3.StringUtils;
import org.gradle.api.Action;
import org.gradle.api.GradleException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * <p>Title:        Godfather1103's Github</p>
 * <p>Copyright:    Copyright (c) 2022</p>
 * <p>Company:      https://github.com/godfather1103</p>
 * 基础的运行
 *
 * @author 作者: Jack Chu E-mail: chuchuanbao@gmail.com
 * <p>
 * 创建时间：2022/3/18 21:31
 * @version 1.0
 * @since 1.0
 */
public abstract class AbstractDockerMojo implements Action<DockerClient> {

    protected DockerPluginExtension ext;

    private final NettyDockerCmdExecFactory factory = new NettyDockerCmdExecFactory();

    public AbstractDockerMojo(DockerPluginExtension ext) {
        this.ext = ext;
    }

    public abstract void initExt(DockerPluginExtension ext);

    private final static Logger logger = LoggerFactory.getLogger(AbstractDockerMojo.class);

    public Logger getLog() {
        return logger;
    }

    public Boolean isSkipDocker() {
        return ext.getSkipDocker().getOrElse(false);
    }


    public int getRetryPushTimeout() {
        return ext.getRetryPushTimeout().getOrElse(10000);
    }

    public int getReadTimeout() {
        return ext.getReadTimeout().getOrElse(0);
    }

    public int getRetryPushCount() {
        return ext.getRetryPushCount().getOrElse(5);
    }

    public boolean isSkipDockerPush() {
        return ext.getSkipDocker().getOrElse(false);
    }

    public void execute() {
        if (isSkipDocker()) {
            getLog().info("Skipping docker goal");
            return;
        }
        // 读取最新的值
        initExt(ext);
        try (DockerClient client = buildDockerClient()) {
            execute(client);
        } catch (Exception e) {
            throw new GradleException("Exception caught", e);
        }
    }

    protected String rawDockerHost() {
        return ext.getDockerHost().getOrNull();
    }

    protected Optional<String> dockerCertificates() {
        String dockerCertPath = ext.getDockerCertPath().getOrNull();
        if (StringUtils.isNotEmpty(dockerCertPath)) {
            return Optional.fromNullable(ext.getDockerCertPath().getOrNull());
        } else {
            return Optional.absent();
        }
    }

    protected AuthConfig makeAuthConfig() {
        // 1、首先从项目的构建脚本中读取
        AuthConfig authConfig = ext.getAuth().getOrNull();
        // 2、没有的话看gradle.properties文件中是否有相关配置
        if (authConfig == null) {
            Object obj = ext.getProject().findProperty("docker.username");
            String username = obj == null ? "" : obj.toString();
            obj = ext.getProject().findProperty("docker.password");
            String password = obj == null ? "" : obj.toString();
            obj = ext.getProject().findProperty("docker.email");
            String email = obj == null ? "" : obj.toString();
            if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(password)) {
                authConfig = new AuthConfig(username, password, email);
            }
        }
        // 3、最后环境看一下环境变量中是否有相关配置
        if (authConfig == null) {
            String username = StringUtils.trimToEmpty(System.getenv("docker.username"));
            String password = StringUtils.trimToEmpty(System.getenv("docker.password"));
            String email = StringUtils.trimToEmpty(System.getenv("docker.email"));
            if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(password)) {
                authConfig = new AuthConfig(username, password, email);
            }
        }
        return authConfig;
    }

    protected void registryAuth(DefaultDockerClientConfig.Builder configBuilder) throws GradleException {
        AuthConfig authConfig = makeAuthConfig();
        if (authConfig != null) {
            final String username = authConfig.getUsername();
            final String password = authConfig.getPassword();
            final String email = authConfig.getEmail();
            if (StringUtils.isNotEmpty(username)) {
                configBuilder.withRegistryUsername(username);
            }
            if (StringUtils.isNotEmpty(email)) {
                configBuilder.withRegistryEmail(email);
            }
            if (StringUtils.isNotEmpty(password)) {
                configBuilder.withRegistryPassword(password);
            }
            configBuilder.withRegistryUrl(serverIdFor());
        } else {
            // settings.xml has no entry for the configured serverId, warn the user
            getLog().warn("No entry found, cannot configure authentication for that registry");
        }
    }

    private String serverIdFor() {
        String serverId = ext.getServerId().getOrNull();
        if (StringUtils.isNotEmpty(serverId)) {
            return serverId;
        }
        String registryUrl = ext.getRegistryUrl().getOrNull();
        if (StringUtils.isNotEmpty(registryUrl)) {
            return ext.getRegistryUrl().getOrNull();
        }
        return "index.docker.io";
    }

    protected DockerClient buildDockerClient() throws GradleException {
        final DefaultDockerClientConfig.Builder configBuilder = DefaultDockerClientConfig.createDefaultConfigBuilder();
        final String dockerHost = rawDockerHost();
        if (StringUtils.isNotEmpty(dockerHost)) {
            configBuilder.withDockerHost(dockerHost);
        }
        final Optional<String> certs = dockerCertificates();
        if (certs.isPresent()) {
            configBuilder.withDockerCertPath(certs.get());
        }
        registryAuth(configBuilder);
        DockerClientConfig config = configBuilder.build();
        return DockerClientImpl.getInstance(config)
                .withDockerCmdExecFactory(factory);
    }
}

File: docker-gradle-plugin-main/src/main/kotlin/com/github/godfather1103/gradle/tasks/AbstractDockerMojo.kt
package com.github.godfather1103.gradle.tasks

import com.github.dockerjava.api.DockerClient
import com.github.dockerjava.core.DefaultDockerClientConfig
import com.github.dockerjava.core.DockerClientConfig
import com.github.dockerjava.core.DockerClientImpl
import com.github.dockerjava.netty.NettyDockerCmdExecFactory
import com.github.godfather1103.gradle.entity.AuthConfig
import com.github.godfather1103.gradle.ext.DockerPluginExtension
import org.apache.commons.lang3.StringUtils
import org.gradle.api.Action
import org.gradle.api.GradleException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*

/**
 * <p>Title:        Godfather1103's Github</p>
 * <p>Copyright:    Copyright (c) 2022</p>
 * <p>Company:      https://github.com/godfather1103</p>
 * 基础的运行
 *
 * @author 作者: Jack Chu E-mail: chuchuanbao@gmail.com
 * <p>
 * 创建时间：2022/3/18 21:31
 * @version 1.0
 * @since 1.0
 */
abstract class AbstractDockerMojo(val ext: DockerPluginExtension) : Action<DockerClient> {

    private val factory = NettyDockerCmdExecFactory()

    private val log = LoggerFactory.getLogger(AbstractDockerMojo::class.java)

    open fun getLog(): Logger {
        return log
    }

    /**
     * 根据扩展配置进行初始化<BR>
     * @author  作者: Jack Chu E-mail: chuchuanbao@gmail.com
     * @date 创建时间：2023/7/27 20:10
     * @param ext 扩展配置
     */
    abstract fun initExt(ext: DockerPluginExtension)

    fun isSkipDocker(): Boolean {
        return ext.skipDocker.getOrElse(false)
    }

    fun getRetryPushTimeout(): Int {
        return ext.retryPushTimeout.getOrElse(10000)
    }

    fun getReadTimeout(): Int {
        return ext.readTimeout.getOrElse(0)
    }

    fun getRetryPushCount(): Int {
        return ext.retryPushCount.getOrElse(5)
    }

    fun isSkipDockerPush(): Boolean {
        return ext.skipDocker.getOrElse(false)
    }

    open fun execute() {
        if (isSkipDocker()) {
            getLog().info("Skipping docker goal")
            return
        }
        // 读取最新值
        initExt(ext)
        try {
            buildDockerClient().use { client -> execute(client) }
        } catch (e: Exception) {
            throw GradleException("Exception caught", e)
        }
    }

    private fun rawDockerHost(): String? {
        return ext.dockerHost.getOrNull()
    }

    private fun dockerCertificates(): Optional<String> {
        val dockerCertPath = ext.dockerCertPath.getOrNull()
        return if (StringUtils.isNotEmpty(dockerCertPath)) {
            Optional.ofNullable(ext.dockerCertPath.getOrNull())
        } else {
            Optional.empty()
        }
    }

    private fun makeAuthConfig(): Optional<AuthConfig> {
        // 1、首先从项目的构建脚本中读取
        var authConfig = ext.auth.getOrNull()
        // 2、没有的话看gradle.properties文件中是否有相关配置
        if (authConfig == null) {
            var obj = ext.project.findProperty("docker.username")
            val username = obj?.toString() ?: ""
            obj = ext.project.findProperty("docker.password")
            val password = obj?.toString() ?: ""
            obj = ext.project.findProperty("docker.email")
            val email = obj?.toString() ?: ""
            if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(password)) {
                authConfig = AuthConfig(username, password, email)
            }
        }
        // 3、最后环境看一下环境变量中是否有相关配置
        if (authConfig == null) {
            val username = StringUtils.trimToEmpty(System.getenv("docker.username"))
            val password = StringUtils.trimToEmpty(System.getenv("docker.password"))
            val email = StringUtils.trimToEmpty(System.getenv("docker.email"))
            if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(password)) {
                authConfig = AuthConfig(username, password, email)
            }
        }
        return Optional.ofNullable(authConfig)
    }

    @Throws(GradleException::class)
    private fun registryAuth(configBuilder: DefaultDockerClientConfig.Builder) {
        val authConfig: Optional<AuthConfig> = makeAuthConfig()
        if (authConfig.isPresent) {
            val username = authConfig.get().username
            val password = authConfig.get().password
            val email = authConfig.get().email
            if (StringUtils.isNotEmpty(username)) {
                configBuilder.withRegistryUsername(username)
            }
            if (StringUtils.isNotEmpty(email)) {
                configBuilder.withRegistryEmail(email)
            }
            if (StringUtils.isNotEmpty(password)) {
                configBuilder.withRegistryPassword(password)
            }
            configBuilder.withRegistryUrl(serverIdFor())
        } else {
            // settings.xml has no entry for the configured serverId, warn the user
            getLog().warn("No entry found, cannot configure authentication for that registry")
        }
    }

    private fun serverIdFor(): String {
        val serverId = ext.serverId.getOrNull()
        if (StringUtils.isNotEmpty(serverId)) {
            return serverId!!
        }
        val registryUrl = ext.registryUrl.getOrNull()
        return if (StringUtils.isNotEmpty(registryUrl)) {
            ext.registryUrl.get()
        } else "index.docker.io"
    }

    @Throws(GradleException::class)
    private fun buildDockerClient(): DockerClient {
        val configBuilder = DefaultDockerClientConfig.createDefaultConfigBuilder()
        val dockerHost = rawDockerHost()
        if (StringUtils.isNotEmpty(dockerHost)) {
            configBuilder.withDockerHost(dockerHost)
        }
        val certs: Optional<String> = dockerCertificates()
        certs.ifPresent {
            configBuilder.withDockerCertPath(it)
        }
        registryAuth(configBuilder)
        val config: DockerClientConfig = configBuilder.build()
        return DockerClientImpl.getInstance(config)
            .withDockerCmdExecFactory(factory)
    }
}